template_loader.go

```diff
@@ -211,3 +212,56 @@ func (h *HttpFilesystemLoader) Get(path string) (io.Reader, error) {
 
 	return h.fs.Open(fullPath)
 }
+
+// EmbededLoader support loading templates from embed.FS
+// embed.FS file names pattern are limited check https://pkg.go.dev/embed#hdr-Directives
+// and https://github.com/golang/go/issues/43854
+type EmbededLoader struct {
+	fs fs.FS
+}
+
+// MustNewEmbededLoader create new instance of EmbededLoader instance
+// and panics if there's any error during instantiation. The parameters
+// are the same like NewEmbededLoader.
+func MustNewEmbededLoader(fs fs.FS) *EmbededLoader {
+	loader, err := NewEmbededLoader(fs)
+	if err != nil {
+		panic(err)
+	}
+	return loader
+}
+
+// NewEmbededLoader creates a new EmbededLoader and allows
+// templates to be loaded from the embed.FS.
+func NewEmbededLoader(fs fs.FS) (*EmbededLoader, error) {
+	if fs == nil {
+		err := errors.New("fs cannot be nil")
+		return nil, err
+	}
+
+	loader := &EmbededLoader{
+		fs: fs,
+	}
+	return loader, nil
+}
+
+// Abs return the join of base and name
+func (e *EmbededLoader) Abs(base, name string) string {
+	return filepath.Join(filepath.Dir(base), name)
+}
+
+// Get returns an io.Reader where the template's content can be read from.
+func (e *EmbededLoader) Get(path string) (reader io.Reader, err error) {
+	f, err := e.fs.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
```

Comment:
Although closing the files is probably best practice, the embedded filesystem does nothing when `Close` is called. We could potentially save quite a bit of performance overhead here, as every file in an embedded filesystem is already loaded and will be handled through garbage collection.

https://cs.opensource.google/go/go/+/refs/tags/go1.17.7:src/embed/embed.go;l=342;drc=refs%2Ftags%2Fgo1.17.7

Reply:
Cool, so we can remove the line, right?

Reply:
I was more thinking there is no need to both open the file, read it and place the contents into a buffer, we should be able to just return the `e.fs.Open` call.